### PR TITLE
[AS-787] 

### PR DESCRIPTION
**Pull request by sagehen03**
It's working for me locally, but I'm looking for guidance on my scala (not idiomatic I'm sure) and on what do about versioning/preserving the old API response.  ~~For now I've replaced the v2 snapshot list end point in place, which is probably wrong, but I didn't want to guess.  It should be easy enough to move things around based on reviewer consensus :crossed_fingers:~~ David looked at the production logs and we've had no callers to the v2 list end point in the last 30 days so seems safe to replace in place. 

Here's some sample JSON responses:

```
//get
{
  "attributes": {
    "instanceName": "terra",
    "snapshot": "6b073ea9-caea-4a64-9cbf-745d17d5a0c8"
  },
  "metadata": {
    "cloningInstructions": "COPY_NOTHING",
    "description": null,
    "name": "testshot",
    "resourceId": "6173b23e-2f87-4fde-9970-747de1a076c1",
    "resourceType": "DATA_REPO_SNAPSHOT",
    "stewardshipType": "REFERENCED",
    "workspaceId": "8f080501-96c3-4aee-8a73-bfab347a2f65"
  }
}

//list 
{ "gcpDataRepoSnapshots": [
  {
    "attributes": {
      "instanceName": "terra",
      "snapshot": "6b073ea9-caea-4a64-9cbf-745d17d5a0c8"
    },
    "metadata": {
      "cloningInstructions": "COPY_NOTHING",
      "description": null,
      "name": "testshot",
      "resourceId": "6173b23e-2f87-4fde-9970-747de1a076c1",
      "resourceType": "DATA_REPO_SNAPSHOT",
      "stewardshipType": "REFERENCED",
      "workspaceId": "8f080501-96c3-4aee-8a73-bfab347a2f65"
    }
  }
]}
```

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/SnapshotApiService.scala
Comment (sagehen03) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/SnapshotApiService.scala:72]: We should delete resourceListToDataReferenceList in the unlikely event we replace v2 enumerate snapshots in place.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala
Comment (sagehen03) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala]: I want to do a references.getResources.map one liner, but it seems like there's some java/scala interop fu that I'm missing as there's no map method on `getResources`

Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala]: `getResources` returns a `java.util.List`, so Scala's `map` isn't available. The java/scala interop fu you're looking for is `import scala.collection.JavaConverters._` … see https://docs.scala-lang.org/overviews/collections/conversions-between-java-and-scala-collections.html for details. Once you've imported that, you have `asScala` available to you, and you can do:
```
    references.getResources.asScala.map{r =>
      val massaged = new DataRepoSnapshotResource
      massaged.setAttributes(r.getResourceAttributes.getGcpDataRepoSnapshot)
      massaged.setMetadata(r.getMetadata)
      massaged
    }
```

this eliminates the use of the `mutable.ListBuffer`, which is good - any mutable classes are un-idiomatic in Scala (of course *_sometimes_* you need one, but it's the exception)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala]: open for debate: I prefer to be explicit about types, so I'd use `Seq.empty[DataRepoSnapshotResource]` instead of `List()`; you could also use `Seq()`. Some prefer the terseness/readability of the shorter `Seq()` but I prefer to see it written out. Scala will do the right thing in either case.

**Review by davidangb — COMMENTED** [review]
Scala commentary inline.

As far as changing the v2 api in place vs. creating a new v3 api, the fallback position is to create a v3 api; that's SOP. In this case, however, these endpoints are only used by a few internal people … I _think_. I suggest the right thing to do is analyze logs and see how often/by whom these are called and make a determination based on that.

**Comment by sagehen03** [comment]
> Scala commentary inline.
> 
> As far as changing the v2 api in place vs. creating a new v3 api, the fallback position is to create a v3 api; that's SOP. In this case, however, these endpoints are only used by a few internal people … I _think_. I suggest the right thing to do is analyze logs and see how often/by whom these are called and make a determination based on that.

I don't think I have access to the prod access logs yet.  How can I check?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:85]: best practice is to make a new case class with a "gcpDataRepoSnapshots" member, then return an instance of that class instead of a `Map`. It's a bit more overhead but it keeps everything type-safe, and other code such as the unit tests doesn't have to be searching through a map for a well-known key.

**Review by davidangb — COMMENTED** [review]
Implementation code looks good to me, modulo my comment about using a case class instead of a Map.

This needs a swagger update - the current response payload as documented is now incorrect.

Before this merges, let's make sure to:
* doublecheck that the v2 API is not used by anyone; I can do this by searching logs
* check with Janki, now that she's back, to ensure PO is ok with changing in place
* ensure we have good release notes, Janki may want to weigh in here
* should we doublecheck with Hornet to make sure this works for them?

**Comment by sagehen03** [comment]
jenkins retest

**Review by davidangb — CHANGES_REQUESTED** [review]
ahhhhh one thing I just thought of: this is going to break nightly automation in `SnapshotApiSpec`, specifically anything that relies on https://github.com/broadinstitute/rawls/blob/83421bec9fb8c00dd35fe01aad34717e3136740f/automation/src/test/scala/org/broadinstitute/dsde/test/api/SnapshotAPISpec.scala#L259.

Note that the nightly automation is unfortunately a separate IntelliJ project / sbt build, so it's a bit of a pain to work with

**Comment by sagehen03** [comment]
> Implementation code looks good to me, modulo my comment about using a case class instead of a Map.
> 
> This needs a swagger update - the current response payload as documented is now incorrect.
> 
> Before this merges, let's make sure to:
> 
>     * doublecheck that the v2 API is not used by anyone; I can do this by searching logs
> 
>     * check with Janki, now that she's back, to ensure PO is ok with changing in place
> 
>     * ensure we have good release notes, Janki may want to weigh in here
> 
>     * should we doublecheck with Hornet to make sure this works for them?

I've updated the swagger and switched to a case class for the response.  Please verify that those changes are ok with our conventions since all the automated things seem to be telling me I'm good (i.e. did I choose the right file to hold the case class? are my swagger changes sufficient?) 

I don't see an obvious release notes file in the rawls codebase.  How do I provide release notes?  

I am still awaiting my prod account; so David or someone else suitable will need to check to verify that the current API is not in use.

I showed Olivia the JSON response at the very beginning of this process and she said it was fine.  I just messaged her about the small change we introduced by putting the list of snapshots under the gcpDataRepoSnapshots key--she says the changes are fine.

I will coordinate with David and Janki over slack.

File: core/src/main/resources/swagger/api-docs.yaml
Comment (davidangb) [inline core/src/main/resources/swagger/api-docs.yaml:4630]: I think `ResourceDescription` is now unused and could be deleted, which in turn would make `ResourceAttributesUnion` unused and deletable. Fine if you'd rather do this in a follow-on PR along with deleting `SnapshotApiService.resourceListToDataReferenceList`

**Review by davidangb — CHANGES_REQUESTED** [review]
these changes look good! will approve once the automation code is fixed too.

Release notes are documented in Jira, fyi

File: core/src/main/resources/swagger/api-docs.yaml
Comment (sagehen03) [reply core/src/main/resources/swagger/api-docs.yaml:4630]: Good point.

**Comment by sagehen03** [comment]
jenkins retest

**Comment by sagehen03** [comment]
jenkins retest

File: automation/project/Dependencies.scala
Comment (davidangb) [inline automation/project/Dependencies.scala]:
```suggestion
    "org.broadinstitute.dsde"       %% "rawls-model"         % "0.1-8223bdce-SNAP"
```

in case the version got lost in Slack threads, here it is

**Review by davidangb — APPROVED** [review]
:+1: let's make sure to keep an eye on the SnapshotAPISpec nightly test against alpha and ensure it passes (it should!)

**Review by MatthewBemis — APPROVED** [review]

**Pull request merged**
